src/EditorFeatures/Core/Implementation/AutomaticCompletion/Sessions/EditorBraceCompletionSession.cs

```diff
@@ -0,0 +1,84 @@
+
+using System.Threading;
+using Microsoft.CodeAnalysis.BraceCompletion;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text.BraceCompletion;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.AutomaticCompletion.Sessions
+{
+    internal class EditorBraceCompletionSession : IEditorBraceCompletionSession
+    {
+        private readonly IBraceCompletionService _braceCompletionService;
+
+        public EditorBraceCompletionSession(IBraceCompletionService braceCompletionService)
+        {
+            _braceCompletionService = braceCompletionService;
+        }
+
+        public BraceCompletionResult? GetChangesAfterReturn(IBraceCompletionSession session, CancellationToken cancellationToken)
+        {
+            var context = GetBraceCompletionContextFromSession(session);
+            if (context == null)
+            {
+                return null;
+            }
+
+            return _braceCompletionService.GetTextChangeAfterReturnAsync(context, supportsVirtualSpace: true, cancellationToken).WaitAndGetResult(cancellationToken);
+        }
+
+        public virtual bool AllowOverType(IBraceCompletionSession session, CancellationToken cancellationToken)
+        {
+            var context = GetBraceCompletionContextFromSession(session);
+            if (context != null)
+            {
+                return _braceCompletionService.AllowOverTypeAsync(context, cancellationToken).WaitAndGetResult(cancellationToken);
+            }
+
+            return true;
+        }
+
+        public BraceCompletionResult? GetBraceCompletion(IBraceCompletionSession session, CancellationToken cancellationToken)
+        {
+            var context = GetBraceCompletionContextFromSession(session);
+            if (context == null)
+            {
+                return null;
+            }
+
+            return _braceCompletionService.GetBraceCompletionAsync(context, cancellationToken).WaitAndGetResult(cancellationToken);
+        }
+
+        public BraceCompletionResult? GetChangesAfterCompletion(IBraceCompletionSession session, CancellationToken cancellationToken)
+        {
+            var context = GetBraceCompletionContextFromSession(session);
+            if (context == null)
+            {
+                return null;
+            }
+
+            return _braceCompletionService.GetTextChangesAfterCompletionAsync(context, cancellationToken).WaitAndGetResult(cancellationToken);
+        }
+
+        public bool IsValidForBraceCompletion(char brace, int openingPosition, Document document, CancellationToken cancellationToken)
+            => _braceCompletionService.IsValidForBraceCompletionAsync(brace, openingPosition, document, cancellationToken).WaitAndGetResult(cancellationToken);
+
+        private static BraceCompletionContext? GetBraceCompletionContextFromSession(IBraceCompletionSession session)
+        {
+            var snapshot = session.SubjectBuffer.CurrentSnapshot;
```

Comment:
1. this type also would benefit from threading guarantees
2. instead of having this type grab from the subject buffer (something inherently mutable), it would be better if the caller grabbed hte immutable pieces that were needed and passed them into this type